File: api-server/app/schemas/backtest.py
"""Backtest Response Schemas (Frozen Dataclasses)"""
from dataclasses import dataclass
from datetime import datetime


@dataclass(frozen=True)
class TradeResponse:
    """거래 응답"""
    id: int
    backtest_run_id: int
    entry_time: datetime
    exit_time: datetime | None
    side: str
    entry_price: float
    exit_price: float | None
    quantity: float
    pnl: float | None
    pnl_pct: float | None
    commission_paid: float | None


@dataclass(frozen=True)
class EquityPointResponse:
    """자산 곡선 포인트 응답"""
    timestamp: datetime
    equity: float
    cash: float
    position_value: float


@dataclass(frozen=True)
class BacktestRunSummaryResponse:
    """백테스트 실행 요약 응답 (목록용)"""
    id: int
    strategy_name: str
    symbol: str
    timeframe: str
    start_date: datetime
    end_date: datetime
    initial_capital: float
    commission: float
    final_capital: float | None
    total_return: float | None
    total_trades: int | None
    win_rate: float | None
    max_drawdown: float | None
    sharpe_ratio: float | None
    created_at: datetime


@dataclass(frozen=True)
class BacktestRunDetailResponse:
    """백테스트 실행 상세 응답"""
    id: int
    strategy_name: str
    symbol: str
    timeframe: str
    start_date: datetime
    end_date: datetime
    initial_capital: float
    commission: float
    final_capital: float | None
    total_return: float | None
    total_trades: int | None
    winning_trades: int | None
    losing_trades: int | None
    win_rate: float | None
    max_drawdown: float | None
    sharpe_ratio: float | None
    parameters: str | None
    created_at: datetime


@dataclass(frozen=True)
class BacktestRunFullResponse:
    """백테스트 실행 전체 응답 (거래 + 자산 곡선)"""
    id: int
    strategy_name: str
    symbol: str
    timeframe: str
    start_date: datetime
    end_date: datetime
    initial_capital: float
    commission: float
    final_capital: float | None
    total_return: float | None
    total_trades: int | None
    winning_trades: int | None
    losing_trades: int | None
    win_rate: float | None
    max_drawdown: float | None
    sharpe_ratio: float | None
    parameters: str | None
    created_at: datetime
    trades: list[TradeResponse]
    equity_curve: list[EquityPointResponse]


@dataclass(frozen=True)
class StatsResponse:
    """통계 응답"""
    total_runs: int
    avg_return: float
    max_return: float
    min_return: float
    strategy_stats: list[dict]


@dataclass(frozen=True)
class AdvancedMetricsResponse:
    """고급 성과 지표 응답"""
    # Risk-adjusted returns
    sortino_ratio: float
    calmar_ratio: float

    # Profitability metrics
    profit_factor: float
    expectancy: float
    win_loss_ratio: float
    recovery_factor: float

    # Streak statistics
    max_consecutive_wins: int
    max_consecutive_losses: int

    # Additional metrics
    net_profit: float
    net_profit_pct: float
    total_commission_paid: float
    avg_trade_duration_hours: float


@dataclass(frozen=True)
class TradeAnalysisResponse:
    """거래 분석 응답"""
    # Holding periods
    avg_holding_hours: float
    median_holding_hours: float
    min_holding_hours: float
    max_holding_hours: float

    # Distribution
    avg_pnl: float
    median_pnl: float
    largest_win: float
    largest_loss: float
    avg_win: float
    avg_loss: float

    # Quality
    small_wins_count: int
    medium_wins_count: int
    large_wins_count: int
    small_losses_count: int
    medium_losses_count: int
    large_losses_count: int

    # MAE/MFE
    avg_mae: float | None
    avg_mfe: float | None
    avg_efficiency: float | None


@dataclass(frozen=True)
class MonteCarloResponse:
    """Monte Carlo 시뮬레이션 결과 응답"""
    n_simulations: int
    initial_capital: float

    # 최종 자본 통계
    final_capital_mean: float
    final_capital_median: float
    final_capital_std: float
    final_capital_5th: float
    final_capital_95th: float
    final_capital_min: float
    final_capital_max: float

    # 수익률 통계
    total_return_mean: float
    total_return_median: float
    total_return_std: float
    total_return_5th: float
    total_return_95th: float

    # MDD 통계
    max_drawdown_mean: float
    max_drawdown_median: float
    max_drawdown_worst: float
    max_drawdown_best: float

    # 샤프 비율 통계
    sharpe_ratio_mean: float
    sharpe_ratio_median: float

    # 리스크 분석
    probability_of_loss: float
    value_at_risk_5: float
    conditional_var_5: float
    confidence_interval_95: tuple[float, float]
    volatility_ratio: float
    risk_grade: str

    # 샘플 데이터
    sample_equity_curves: list[list[float]]
    all_final_capitals: list[float]


@dataclass(frozen=True)
class CostStressResponse:
    """Cost Stress Test 결과 응답"""
    # Base case
    base_sharpe: float
    base_cagr: float
    base_total_return: float
    base_max_drawdown: float

    # Commission 2x stress
    comm_2x_sharpe: float
    comm_2x_cagr: float
    comm_2x_total_return: float
    comm_2x_sharpe_delta: float
    comm_2x_sharpe_delta_pct: float

    # Slippage +1 tick stress
    slip_1tick_sharpe: float
    slip_1tick_cagr: float
    slip_1tick_total_return: float
    slip_1tick_cagr_delta: float
    slip_1tick_cagr_delta_pct: float

    # Execution delay (1 bar)
    delay_1bar_sharpe: float
    delay_1bar_cagr: float
    delay_1bar_total_return: float
    delay_1bar_max_drawdown: float
    delay_1bar_total_return_delta: float
    delay_1bar_total_return_delta_pct: float

    # Risk assessment
    passes_stress_test: bool
    failure_reasons: list[str]
    risk_grade: str

File: web-dashboard/src/types/backtest.ts
// TypeScript interfaces for backtest data
// Mirrors backend schemas from api-server/app/schemas/backtest.py

export interface BacktestRunSummary {
  id: number;
  strategy_name: string;
  symbol: string;
  timeframe: string;
  start_date: string;
  end_date: string;
  initial_capital: number;
  commission: number;
  final_capital: number | null;
  total_return: number | null;
  total_trades: number | null;
  win_rate: number | null;
  max_drawdown: number | null;
  sharpe_ratio: number | null;
  created_at: string;
}

export interface BacktestRunDetail extends BacktestRunSummary {
  winning_trades: number | null;
  losing_trades: number | null;
  parameters: string | null;
}

export interface Trade {
  id: number;
  backtest_run_id: number;
  entry_time: string;
  exit_time: string | null;
  side: 'long' | 'short';
  entry_price: number;
  exit_price: number | null;
  quantity: number;
  pnl: number | null;
  pnl_pct: number | null;
  commission_paid: number | null;
}

export interface EquityPoint {
  timestamp: string;
  equity: number;
  cash: number;
  position_value: number;
}

export interface AdvancedMetrics {
  sortino_ratio: number;
  calmar_ratio: number;
  profit_factor: number;
  expectancy: number;
  win_loss_ratio: number;
  recovery_factor: number;
  max_consecutive_wins: number;
  max_consecutive_losses: number;
  net_profit: number;
  net_profit_pct: number;
  total_commission_paid: number;
  avg_trade_duration_hours: number;
}

export interface TradeAnalysis {
  avg_holding_hours: number;
  median_holding_hours: number;
  min_holding_hours: number;
  max_holding_hours: number;
  avg_pnl: number;
  median_pnl: number;
  largest_win: number;
  largest_loss: number;
  avg_win: number;
  avg_loss: number;
  small_wins_count: number;
  medium_wins_count: number;
  large_wins_count: number;
  small_losses_count: number;
  medium_losses_count: number;
  large_losses_count: number;
  avg_mae: number | null;
  avg_mfe: number | null;
  avg_efficiency: number | null;
}

export interface MonteCarloResult {
  n_simulations: number;
  initial_capital: number;

  // 최종 자본 통계
  final_capital_mean: number;
  final_capital_median: number;
  final_capital_std: number;
  final_capital_5th: number;
  final_capital_95th: number;
  final_capital_min: number;
  final_capital_max: number;

  // 수익률 통계
  total_return_mean: number;
  total_return_median: number;
  total_return_std: number;
  total_return_5th: number;
  total_return_95th: number;

  // MDD 통계
  max_drawdown_mean: number;
  max_drawdown_median: number;
  max_drawdown_worst: number;
  max_drawdown_best: number;

  // 샤프 비율 통계
  sharpe_ratio_mean: number;
  sharpe_ratio_median: number;

  // 리스크 분석
  probability_of_loss: number;
  value_at_risk_5: number;
  conditional_var_5: number;
  confidence_interval_95: [number, number];
  volatility_ratio: number;
  risk_grade: string;

  // 샘플 데이터
  sample_equity_curves: number[][];
  all_final_capitals: number[];
}

export interface CostStressResult {
  // Base case
  base_sharpe: number;
  base_cagr: number;
  base_total_return: number;
  base_max_drawdown: number;

  // Commission 2x stress
  comm_2x_sharpe: number;
  comm_2x_cagr: number;
  comm_2x_total_return: number;
  comm_2x_sharpe_delta: number;
  comm_2x_sharpe_delta_pct: number;

  // Slippage +1 tick stress
  slip_1tick_sharpe: number;
  slip_1tick_cagr: number;
  slip_1tick_total_return: number;
  slip_1tick_cagr_delta: number;
  slip_1tick_cagr_delta_pct: number;

  // Execution delay (1 bar)
  delay_1bar_sharpe: number;
  delay_1bar_cagr: number;
  delay_1bar_total_return: number;
  delay_1bar_max_drawdown: number;
  delay_1bar_total_return_delta: number;
  delay_1bar_total_return_delta_pct: number;

  // Risk assessment
  passes_stress_test: boolean;
  failure_reasons: string[];
  risk_grade: string;
}

export interface BacktestRunFull extends BacktestRunDetail {
  trades: Trade[];
  equity_curve: EquityPoint[];
}

// Computed analytics interfaces
export interface DrawdownPoint {
  timestamp: string;
  drawdown_pct: number;
  peak_equity: number;
}

export interface RollingMetrics {
  timestamp: string;
  rolling_sharpe: number;
  rolling_volatility: number;
  rolling_return: number;
}
